<a href="https://colab.research.google.com/github/ikwak2/StatBook/blob/main/ch11/11_6_python_prac.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 11.6 Python 연습

## <예 11-1>에 대한 카이제곱 적합도 검정 과정
### 자료의 입력


In [10]:
import numpy as np
p = np.array([713, 215, 22, 52])
x = np.array([690, 232, 234, 44])

print("p:", p)
print("x:", x)

p: [713 215  22  52]
x: [690 232 234  44]


여기서 p는 기대도수, x는 관측도수를 나타낸다.

### 카이제곱 검정


In [8]:
from scipy.stats import chisquare

chisquare(x, f_exp=p)


ValueError: ignored

R에서 기본으로 제공하는 명령어 chisq.test를 이용하여 회귀분석을 실시한다. 명령문에서 괄호 안의 첫 번째는 관측도수, 두 번째는 기대도수이다. 세 번째 옵션 rescale.p=T는 두 번째 입력변수 p를 관측도수로 사용할 것인지의 여부이다. rescale=F를 입력할 경우 p에는 관측도수가 아닌 비율이 들어가야 한다. (이 예제에서는 c(9/16,3/16,3/16,1/16)) 

In [ ]:
import pandas as pd

ex10 = pd.read_csv("ex10_2.csv")
print(ex10)

    value method
0      65      A
1      87      A
2      73      A
3      79      A
4      81      A
5      69      A
6      75      B
7      69      B
8      83      B
9      81      B
10     72      B
11     79      B
12     90      B
13     59      C
14     78      C
15     67      C
16     62      C
17     83      C
18     76      C
19     94      D
20     89      D
21     80      D
22     88      D


pd.read_csv 명령어를 이용하여 ex10_2.csv 파일을 ex10에 저장한다. 첫 번째 변수 value는 시험성적, 두 번째 변수 method는 교육방법이다. 변수명 및 데이터경로에 한글이 있을 경우 오류가 생길 수 있으므로, 가능하면 영문을 사용하는 것이 좋다.  

### 일원분류 분산분석

In [ ]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

ex10_aov = ols('value ~ method', data=ex10).fit() ## anova를 위한 모형 적합

print(sm.stats.anova_lm(ex10_aov))

            df       sum_sq     mean_sq         F    PR(>F)
method     3.0   712.586439  237.528813  3.771461  0.028041
Residual  19.0  1196.630952   62.980576       NaN       NaN


python의 statsmodels 라이브러리에서 제공하는 명령어 anova_lm를 이용하여 분산분석을 실시한다. 명령문에서 statsmodels.formula.api.ols 를 통해 분산분석을 위한 식을 입력한다. ‘~’기호 앞의 변수가 종속변수, 기호 뒤의 변수는 요인이 된다. 위의 식은 교육방법(A, B, C, D)에 따른 시험성적의 평균 차이가 있는가를 알아보는 분석이다. 이 경우 사용하는 자료 ex10 안에 식에 사용되는 변수 value와 method가 포함되어 있어야 한다.

### 결과 해석

$\mu_1 = \mu_2 = \mu_3 = \mu_4 $를 각각의 교육방법에서 모집단의 평균성적이라 할 때, 검정하고자 하는 가설은 다음과 같다.

$H_0: \mu_1 = \mu_2 = \mu_3 = \mu_4$ vs $H_1 : H_0$ 는 사실이 아니다.

P값이 0.02804이므로 5% 유의수준 하에서 귀무가설은 기각된다. 즉, 주어진 자료에 의할 때 5% 유의수준 하에서 4가지 교육방법의 효과가 동일하다고 볼 수 없다.

## <예제 10-4> 자료에 대한 분산분석 과정
### 자료 입력

In [ ]:

ex10_2 = pd.read_csv("ex10_4.csv")
ex10_2

value treat block
0     1.7     A     a
1     1.5     A     b
2     0.1     A     c
3     0.6     A     d
4     3.4     B     a
5     2.6     B     b
6     2.3     B     c
7     2.2     B     d
8     2.3     C     a
9     2.1     C     b
10    0.8     C     c
11    1.6     C     d

read.csv 명령어를 이용하여 ex10_4.csv 파일을 ex10_2 에 저장한다. 첫 번째 변수 value는 반응시간, 두 번째 변수 treat는 자극의 종류, 세 번째 변수 block은 사람으로 블록 효과이다. 

In [ ]:
import statsmodels
import statsmodels.api as sm

ex_10_two_aov = statsmodels.formula.api.ols(formula="value ~ block + treat", data=ex10_2).fit()
ex_10_two_aov_anova = sm.stats.anova_lm(ex_10_two_aov)

ex_10_two_aov_anova

df    sum_sq   mean_sq          F    PR(>F)
block     3.0  3.480000  1.160000  15.296703  0.003232
treat     2.0  5.471667  2.735833  36.076923  0.000452
Residual  6.0  0.455000  0.075833        NaN       NaN

statsmodels 라이브러리를 활용하여 명령어 statsmodels.api.stats.anova_lm를 이용하여 분산분석을 실시한다. statsmodels.formula.api.ols 명령문을 이용하여 분산분석을 위한 식을 입력한다. ‘~’기호 앞의 변수는 종속변수, 기호 뒤의 변수는 요인이 된다. 두 가지 이상의 요인이 필요한 경우 ‘+’ 기호를 사용한다. 위의 코드는 블록(block : 사람 1, 사람 2, 사람 3, 사람 4) 효과와 자극의 종류(treat : A, B, C)에 따른 반응시간의 처리 효과를 분석하는 코드로, 사용하는 자료 ex10_2 안에 식에 사용되는 변수 value와 block, treat가 포함되어 있어야 한다.

### 결과 해석

$\beta_1 = \beta_2 = \beta_3 = \beta_4$를 각각의 사람(여기서는 block 변수)에 따른 반응시간의 모평균이라 할 때, 검정하고자 하는 가설은 다음과 같다.  
$H_0: \beta_1 = \beta_2 = \beta_3 = \beta_4$ vs $H_1: H_0 $은 사실이 아니다. 

위의 분석결과에서 block의 효과에 대한 P값이 0.0032316으로 0.05보다 작으므로 귀무가설은 기각된다. 즉, 사람에 따른 반응시간의 기댓값이 동일하다고 할 수 없다.